File: celesta/score/aggregate/testAggregate.py
# coding=UTF-8

from java.sql import Timestamp
from java.time import LocalDateTime

from aggregate._aggregate_orm import countConditionLessCursor, countGetDateCondCursor\
    , viewCountCondLessCursor, viewCountGetDateCondCursor, tableSumOneFieldCursor\
    , viewSumOneFieldCursor, sumFieldAndNumberCursor, viewSumTwoNumbersCursor\
    , tableSumTwoFieldsCursor, viewSumTwoFieldsCursor, tableMinMaxCursor\
    , viewMinOneFieldCursor, viewMaxOneFieldCursor, viewMinTwoFieldsCursor, viewMaxTwoFieldsCursor\
    , tableGroupByCursor, viewGroupByAggregateCursor, viewGroupByCursor, viewCountMinMaxCursor
from ru.curs.celesta.unit import TestClass, CelestaTestCase


@TestClass
class TestAggregate(CelestaTestCase):

    def test_count_without_condition(self):
        tableCursor = countConditionLessCursor(self.context)
        tableCursor.deleteAll()

        viewCursor = viewCountCondLessCursor(self.context)
        self.assertEquals(1, viewCursor.count())
        viewCursor.first()
        self.assertEquals(0, viewCursor.c)

        tableCursor.insert()
        tableCursor.clear()
        tableCursor.insert()

        self.assertEquals(1, viewCursor.count())
        viewCursor.first()
        self.assertEquals(2, viewCursor.c)


    def test_count_with_getdate_condition(self):
        tableCursor = countGetDateCondCursor(self.context)
        tableCursor.deleteAll()

        viewCursor = viewCountGetDateCondCursor(self.context)
        viewCursor.first()
        self.assertEquals(0, viewCursor.c)

        tableCursor.insert()
        tableCursor.clear()
        tableCursor.date = Timestamp.valueOf(LocalDateTime.now().minusSeconds(1))
        tableCursor.insert()

        viewCursor.first()
        self.assertEquals(0, viewCursor.c)

        tableCursor.clear()
        tableCursor.date = Timestamp.valueOf(LocalDateTime.now().plusDays(1))
        tableCursor.insert()

        viewCursor.first()
        self.assertEquals(1, viewCursor.c)


    def test_sum_one_field(self):
        tableOneFieldCursor = tableSumOneFieldCursor(self.context)
        tableOneFieldCursor.deleteAll()

        viewOneFieldCursor = viewSumOneFieldCursor(self.context)
        viewOneFieldAndNumberCursor= sumFieldAndNumberCursor(self.context)
        viewTwoNumbersCursor = viewSumTwoNumbersCursor(self.context)

        self.assertEquals(1, viewOneFieldCursor.count())
        self.assertEquals(1, viewOneFieldAndNumberCursor.count())
        self.assertEquals(1, viewTwoNumbersCursor.count())

        viewOneFieldCursor.first()
        viewOneFieldAndNumberCursor.first()
        viewTwoNumbersCursor.first()
        self.assertEquals(None, viewOneFieldCursor.s)
        self.assertEquals(None, viewOneFieldAndNumberCursor.s)
        self.assertEquals(None, viewTwoNumbersCursor.s)

        tableOneFieldCursor.f = 4
        tableOneFieldCursor.insert()

        viewOneFieldCursor.first()
        viewOneFieldAndNumberCursor.first()
        viewTwoNumbersCursor.first()
        self.assertEquals(4, viewOneFieldCursor.s)
        self.assertEquals(5, viewOneFieldAndNumberCursor.s)
        self.assertEquals(3, viewTwoNumbersCursor.s)


    def test_sum_two_fields(self):
        tableTwoFieldsCursor = tableSumTwoFieldsCursor(self.context)
        tableTwoFieldsCursor.deleteAll()

        viewTwoFieldsCursor = viewSumTwoFieldsCursor(self.context)

        self.assertEquals(1, viewTwoFieldsCursor.count())

        viewTwoFieldsCursor.first()
        self.assertEquals(None, viewTwoFieldsCursor.s)

        tableTwoFieldsCursor.f1 = 2
        tableTwoFieldsCursor.insert()
        tableTwoFieldsCursor.clear()

        viewTwoFieldsCursor.first()
        self.assertEquals(None, viewTwoFieldsCursor.s)

        tableTwoFieldsCursor.f2 = 2
        tableTwoFieldsCursor.insert()
        tableTwoFieldsCursor.clear()

        viewTwoFieldsCursor.first()
        self.assertEquals(None, viewTwoFieldsCursor.s)

        tableTwoFieldsCursor.f1 = 2
        tableTwoFieldsCursor.f2 = 3
        tableTwoFieldsCursor.insert()
        tableTwoFieldsCursor.clear()

        viewTwoFieldsCursor.first()
        self.assertEquals(5, viewTwoFieldsCursor.s)


    def test_min_and_max_one_field(self):
      tableCur = tableMinMaxCursor(self.context)
      tableCur.deleteAll()

      viewMinOneFieldCur = viewMinOneFieldCursor(self.context)
      viewMaxOneFieldCur = viewMaxOneFieldCursor(self.context)
      viewMinTwoFieldsCur = viewMinTwoFieldsCursor(self.context)
      viewMaxTwoFieldsCur = viewMaxTwoFieldsCursor(self.context)
      viewCountMinMaxCur = viewCountMinMaxCursor(self.context)

      viewMinOneFieldCur.first()
      viewMaxOneFieldCur.first()
      viewMinTwoFieldsCur.first()
      viewMaxTwoFieldsCur.first()
      viewCountMinMaxCur.first()
      self.assertEquals(1, viewMinOneFieldCur.count())
      self.assertEquals(1, viewMaxOneFieldCur.count())
      self.assertEquals(1, viewMinTwoFieldsCur.count())
      self.assertEquals(1, viewMaxTwoFieldsCur.count())
      self.assertEquals(1, viewCountMinMaxCur.count())
      self.assertEquals(None, viewMinOneFieldCur.m)
      self.assertEquals(None, viewMaxOneFieldCur.m)
      self.assertEquals(None, viewMinTwoFieldsCur.m)
      self.assertEquals(None, viewMaxTwoFieldsCur.m)
      self.assertEquals(0, viewCountMinMaxCur.countv)
      self.assertEquals(None, viewCountMinMaxCur.maxv)
      self.assertEquals(None, viewCountMinMaxCur.minv)

      tableCur.f1 = 1
      tableCur.f2 = 5
      tableCur.insert()
      tableCur.clear()

      tableCur.f1 = 5
      tableCur.f2 = 2
      tableCur.insert()
      tableCur.clear()

      viewMinOneFieldCur.first()
      viewMaxOneFieldCur.first()
      viewMinTwoFieldsCur.first()
      viewMaxTwoFieldsCur.first()
      viewCountMinMaxCur.first()

      self.assertEquals(1, viewMinOneFieldCur.m)
      self.assertEquals(5, viewMaxOneFieldCur.m)
      self.assertEquals(6, viewMinTwoFieldsCur.m)
      self.assertEquals(7, viewMaxTwoFieldsCur.m)
      self.assertEquals(2, viewCountMinMaxCur.countv)
      self.assertEquals(5, viewCountMinMaxCur.maxv)
      self.assertEquals(2, viewCountMinMaxCur.minv)


    def testGroupBy(self):
        tableCursor = tableGroupByCursor(self.context)
        tableCursor.deleteAll()

        viewGroupByCur = viewGroupByCursor(self.context)
        viewAggregateCursor = viewGroupByAggregateCursor(self.context)

        self.assertEquals(0, viewAggregateCursor.count())

        name1 = "A"
        name2 = "B"

        tableCursor.name = name1
        tableCursor.cost = 100
        tableCursor.insert()
        tableCursor.clear()

        viewGroupByCur.first()
        self.assertEquals(name1, viewGroupByCur.name)
        self.assertEquals(100, viewGroupByCur.cost)

        tableCursor.name = name1
        tableCursor.cost = 150
        tableCursor.insert()
        tableCursor.clear()
        tableCursor.name = name2
        tableCursor.cost = 50
        tableCursor.insert()
        tableCursor.clear()

        self.assertEquals(2, viewAggregateCursor.count())
        viewAggregateCursor.first()
        self.assertEquals(name1, viewAggregateCursor.name)
        self.assertEquals(250, viewAggregateCursor.s)

        viewAggregateCursor.next()
        self.assertEquals(name2, viewAggregateCursor.name)
        self.assertEquals(50, viewAggregateCursor.s)

File: celesta-test/src/test/java/ru/curs/celesta/script/TestAggregate.java
package ru.curs.celesta.script;

import aggregate.*;
import org.junit.jupiter.api.TestTemplate;
import ru.curs.celesta.CallContext;

import java.sql.Timestamp;
import java.time.LocalDateTime;
import java.math.BigDecimal;

import static org.junit.jupiter.api.Assertions.*;


class TestAggregate implements ScriptTest {

    @TestTemplate
    void test_count_without_condition(CallContext context) {
        CountConditionLessCursor tableCursor = new CountConditionLessCursor(context);
        tableCursor.deleteAll();

        ViewCountCondLessCursor viewCursor = new ViewCountCondLessCursor(context);
        assertEquals(1, viewCursor.count());
        viewCursor.first();
        assertEquals(0, viewCursor.getC().intValue());

        tableCursor.insert();
        tableCursor.clear();
        tableCursor.insert();

        assertEquals(1, viewCursor.count());
        viewCursor.first();
        assertEquals(2, viewCursor.getC().intValue());


    }

    @TestTemplate
    void test_count_with_getdate_condition(CallContext context) {
        CountGetDateCondCursor tableCursor = new CountGetDateCondCursor(context);
        tableCursor.deleteAll();

        ViewCountGetDateCondCursor viewCursor = new ViewCountGetDateCondCursor(context);
        viewCursor.first();
        assertEquals(0, viewCursor.getC().intValue());

        tableCursor.insert();
        tableCursor.clear();
        tableCursor.setDate(Timestamp.valueOf(LocalDateTime.now().minusSeconds(2)));
        tableCursor.insert();

        viewCursor.first();
        assertEquals(0, viewCursor.getC().intValue());

        tableCursor.clear();
        tableCursor.setDate(Timestamp.valueOf(LocalDateTime.now().plusDays(1)));
        tableCursor.insert();

        viewCursor.first();
        assertEquals(1, viewCursor.getC().intValue());
    }

    @TestTemplate
    void test_sum_one_field(CallContext context) {
        TableSumOneFieldCursor tableOneFieldCursor = new TableSumOneFieldCursor(context);
        tableOneFieldCursor.deleteAll();

        ViewSumOneFieldCursor viewOneFieldCursor = new ViewSumOneFieldCursor(context);
        SumFieldAndNumberCursor viewOneFieldAndNumberCursor = new SumFieldAndNumberCursor(context);
        ViewSumTwoNumbersCursor viewTwoNumbersCursor = new ViewSumTwoNumbersCursor(context);

        assertEquals(1, viewOneFieldCursor.count());
        assertEquals(1, viewOneFieldAndNumberCursor.count());
        assertEquals(1, viewTwoNumbersCursor.count());

        viewOneFieldCursor.first();
        viewOneFieldAndNumberCursor.first();
        viewTwoNumbersCursor.first();
        assertEquals(null, viewOneFieldCursor.getS());
        assertEquals(null, viewOneFieldAndNumberCursor.getS());
        assertEquals(null, viewTwoNumbersCursor.getS());

        tableOneFieldCursor.setF(4);
        tableOneFieldCursor.insert();

        viewOneFieldCursor.first();
        viewOneFieldAndNumberCursor.first();
        viewTwoNumbersCursor.first();
        assertEquals(4, viewOneFieldCursor.getS().intValue());
        assertEquals(5, viewOneFieldAndNumberCursor.getS().intValue());
        assertEquals(3, viewTwoNumbersCursor.getS().intValue());


    }

    @TestTemplate
    void test_sum_two_fields(CallContext context) {
        TableSumTwoFieldsCursor tableTwoFieldsCursor = new TableSumTwoFieldsCursor(context);
        tableTwoFieldsCursor.deleteAll();

        ViewSumTwoFieldsCursor viewTwoFieldsCursor = new ViewSumTwoFieldsCursor(context);

        assertEquals(1, viewTwoFieldsCursor.count());

        viewTwoFieldsCursor.first();
        assertEquals(null, viewTwoFieldsCursor.getS());

        tableTwoFieldsCursor.setF1(2);
        tableTwoFieldsCursor.insert();
        tableTwoFieldsCursor.clear();

        viewTwoFieldsCursor.first();
        assertEquals(null, viewTwoFieldsCursor.getS());

        tableTwoFieldsCursor.setF2(2);
        tableTwoFieldsCursor.insert();
        tableTwoFieldsCursor.clear();

        viewTwoFieldsCursor.first();
        assertEquals(null, viewTwoFieldsCursor.getS());

        tableTwoFieldsCursor.setF1(2);
        tableTwoFieldsCursor.setF2(3);
        tableTwoFieldsCursor.insert();
        tableTwoFieldsCursor.clear();

        viewTwoFieldsCursor.first();
        assertEquals(5, viewTwoFieldsCursor.getS().intValue());


    }

    @TestTemplate
    void test_min_and_max_one_field(CallContext context) {
        TableMinMaxCursor tableCur = new TableMinMaxCursor(context);
        tableCur.deleteAll();

        ViewMinOneFieldCursor viewMinOneFieldCur = new ViewMinOneFieldCursor(context);
        ViewMaxOneFieldCursor viewMaxOneFieldCur = new ViewMaxOneFieldCursor(context);
        ViewMinTwoFieldsCursor viewMinTwoFieldsCur = new ViewMinTwoFieldsCursor(context);
        ViewMaxTwoFieldsCursor viewMaxTwoFieldsCur = new ViewMaxTwoFieldsCursor(context);
        ViewCountMinMaxCursor viewCountMinMaxCur = new ViewCountMinMaxCursor(context);

        viewMinOneFieldCur.first();
        viewMaxOneFieldCur.first();
        viewMinTwoFieldsCur.first();
        viewMaxTwoFieldsCur.first();
        viewCountMinMaxCur.first();
        assertEquals(1, viewMinOneFieldCur.count());
        assertEquals(1, viewMaxOneFieldCur.count());
        assertEquals(1, viewMinTwoFieldsCur.count());
        assertEquals(1, viewMaxTwoFieldsCur.count());
        assertEquals(1, viewCountMinMaxCur.count());
        assertEquals(null, viewMinOneFieldCur.getM());
        assertEquals(null, viewMaxOneFieldCur.getM());
        assertEquals(null, viewMinTwoFieldsCur.getM());
        assertEquals(null, viewMaxTwoFieldsCur.getM());
        assertEquals(0, viewCountMinMaxCur.getCountv().intValue());
        assertEquals(null, viewCountMinMaxCur.getMaxv());
        assertEquals(null, viewCountMinMaxCur.getMinv());

        tableCur.setF1(1);
        tableCur.setF2(5);
        tableCur.insert();
        tableCur.clear();

        tableCur.setF1(5);
        tableCur.setF2(2);
        tableCur.insert();
        tableCur.clear();

        viewMinOneFieldCur.first();
        viewMaxOneFieldCur.first();
        viewMinTwoFieldsCur.first();
        viewMaxTwoFieldsCur.first();
        viewCountMinMaxCur.first();

        assertEquals(1, viewMinOneFieldCur.getM().intValue());
        assertEquals(5, viewMaxOneFieldCur.getM().intValue());
        assertEquals(6, viewMinTwoFieldsCur.getM().intValue());
        assertEquals(7, viewMaxTwoFieldsCur.getM().intValue());
        assertEquals(2, viewCountMinMaxCur.getCountv().intValue());
        assertEquals(5, viewCountMinMaxCur.getMaxv().intValue());
        assertEquals(2, viewCountMinMaxCur.getMinv().intValue());


    }

    @TestTemplate
    void testGroupBy(CallContext context) {
        TableGroupByCursor tableCursor = new TableGroupByCursor(context);
        tableCursor.deleteAll();

        ViewGroupByCursor viewGroupByCur = new ViewGroupByCursor(context);
        ViewGroupByAggregateCursor viewAggregateCursor = new ViewGroupByAggregateCursor(context);

        assertEquals(0, viewAggregateCursor.count());

        String name1 = "A";
        String name2 = "B";

        tableCursor.setName(name1);
        tableCursor.setCost(100);
        tableCursor.insert();
        tableCursor.clear();

        viewGroupByCur.first();
        assertEquals(name1, viewGroupByCur.getName());
        assertEquals(100, viewGroupByCur.getCost().intValue());

        tableCursor.setName(name1);
        tableCursor.setCost(150);
        tableCursor.insert();
        tableCursor.clear();
        tableCursor.setName(name2);
        tableCursor.setCost(50);
        tableCursor.insert();
        tableCursor.clear();

        assertEquals(2, viewAggregateCursor.count());
        viewAggregateCursor.first();
        assertEquals(name1, viewAggregateCursor.getName());
        assertEquals(250, viewAggregateCursor.getS().intValue());

        viewAggregateCursor.next();
        assertEquals(name2, viewAggregateCursor.getName());
        assertEquals(50, viewAggregateCursor.getS().intValue());

    }

    @TestTemplate
    void testSumOfDecimal(CallContext context) {
        TWithDecimalCursor t = new TWithDecimalCursor(context);
        ViewWithDecimalCursor v = new ViewWithDecimalCursor(context);

        t.insert();
        t.clear();
        t.insert();

        v.first();

        assertEquals(new BigDecimal("48.02"), v.getF1().stripTrailingZeros());
        assertEquals(new BigDecimal("2.0002"), v.getF2().stripTrailingZeros());
        assertEquals(new BigDecimal("50.0202"), v.getF12().stripTrailingZeros());
    }
}
